app/src/main/java/com/x8bit/bitwarden/data/autofill/fido2/processor/Fido2ProviderProcessorImpl.kt

```diff
@@ -124,10 +141,118 @@ class Fido2ProviderProcessorImpl(
+                ),
+            )
+            return
+        }
+
+        // Otherwise, find all matching credentials from the current vault.
+        try {
+            runBlocking {
+                val credentialEntries = getMatchingFido2CredentialEntries(
+                    userId = userState.activeUserId,
+                    request = request,
+                )
+
+                callback.onResult(
+                    BeginGetCredentialResponse(
+                        credentialEntries = credentialEntries,
+                    ),
+                )
+            }
+        } catch (e: GetCredentialUnknownException) {
+            callback.onError(e)
+        }
     }
 
+    @Throws
+    private suspend fun getMatchingFido2CredentialEntries(
+        userId: String,
+        request: BeginGetCredentialRequest,
+    ): List<CredentialEntry> =
+        request
+            .beginGetCredentialOptions
+            .flatMap { option ->
+                if (option is BeginGetPublicKeyCredentialOption) {
+                    val requestRpId = Json
+                        .decodeFromStringOrNull<PublicKeyCredentialRequestOptions>(
+                            option.requestJson,
+                        )
+                        ?.relayingPartyId
+                        ?: throw GetCredentialUnknownException("Invalid data.")
+
+                    vaultRepository
+                        .ciphersStateFlow
+                        .value
+                        .data
+                        ?.filter { it.isActiveWithFido2Credentials }
+                        ?.let {
+                            vaultSdkSource.decryptFido2CredentialAutofillViews(
+                                userId = userId,
+                                cipherViews = it.toTypedArray(),
+                            )
+                        }
+                        ?.fold(
+                            onSuccess = { fido2CredentialAutofillViews ->
+                                fido2CredentialAutofillViews
+                                    .filter { it.rpId == requestRpId }
+                                    .toCredentialEntries(option)
+                            },
+                            onFailure = {
+                                throw GetCredentialUnknownException("Error decrypting credentials.")
+                            },
+                        )
+                        .orEmpty()
+                } else {
+                    throw GetCredentialUnknownException("Unsupported option.")
+                }
+            }
+
+    private fun List<Fido2CredentialAutofillView>.toCredentialEntries(
+        option: BeginGetPublicKeyCredentialOption,
+    ): List<CredentialEntry> =
+        this
+            .map {
+                PublicKeyCredentialEntry
+                    .Builder(
+                        context = context,
+                        username = it.rpId,
```

Comment:
This seems suspect. rpId is the relying part identifier, which is often a URI. It's odd we're assigning that to a `username` property. I think we should be using `Fido2CredentialAutofillView.userNameForUi` here instead.

Reply:
Ah ok that's good to know. What should we do if the `userNameforUi` property is null? It looks like it's optional in the `Fido2CredentialAutofillView`

Reply:
@coroiu said the SDK will return "None" when a username cannot be derived. In those cases we need to display "No Username". The SDK can't provide us with the translated text so we'll have to map it to an appropriate resource on our end. If we do ever receive a null value we can use the same "No Username" resource.